### PR TITLE
Add option for specifying plugins directories

### LINKED ISSUE #3260: Add `--plugin-dir` and/or `--no-plugins` options
### Checklist

- [X] I'm reporting a feature request
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2022.03.08.1**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Description

I've got a directory: `${HOME}/bin/ytdlp_plugins/`

On the command line, if I type `yt<TAB>` in the command position, it gives me options to expand for both `yt-dlp` and `ytdlp_plugins`. This is annoying, since I'm not running anything from ytdlp_plugins directly from the command line.

Request: Can there be an option to put the `ytdlp_plugins` directory in an alternate location? So it doesn't trip up tab-completion on the command line?

Ideally, I'd like to have a `--plugins-dir` (or similar) option, so I can put the plugins directly **not** in my executables path.

### PR DESCRIPTION
**Pull request by imranh2**
Closes #3260

**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Lets you specify `--plugin-locations` 1 or more times and then goes to each specified location and looks for `<package name>/yt_dlp_plugins/` as normal. Via CLI or config file.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by Grub4K** [comment]
I think this should be implemented using an environment variable, similar to `YTDLP_NO_PLUGINS`. The commandline flag will then simply modify the env var.

The points up for discussion and my suggestions:
- Should multiple inputs be merged (yes, env var appends last)
- Should we have a sentinel to reset all (yes, `none`)
- What should the name for the option be (`--plugin-dir`)
- What should the name for the env var be (`YTDLP_PLUGIN_DIRS`)

**Comment by coletdjnz** [comment]
> I think this should be implemented using an environment variable, similar to `YTDLP_NO_PLUGINS`. The commandline flag will then simply modify the env var.
> 
> The points up for discussion and my suggestions:
> - Should multiple inputs be merged (yes, env var appends last)
> - Should we have a sentinel to reset all (yes, `none`)
> - What should the name for the option be (`--plugin-dir`)
> - What should the name for the env var be (`YTDLP_PLUGIN_DIRS`)

We should avoid modifying an env var imo. Only reas from one.

Otherwise that state will persist

**Review by coletdjnz — APPROVED** [review]
it's dirty but it works for now

only public part of this is --plugin-locations opt which I'm happy with as we will reuse it when we implement the globals system

api users can use sys.path

File: yt_dlp/options.py [hunk 1]
```diff
@@ -408,6 +408,10 @@ def _alias_callback(option, opt_str, value, parser, opts, nargs):
         help=(
             'Location of the main configuration file; either the path to the config or its containing directory '
             '("-" for stdin). Can be used multiple times and inside other configuration files'))
+    general.add_option(
+        '--plugin-locations',
+        dest='plugin_locations', metavar='PATH', action='append',
```
Comment (Grub4K) [inline yt_dlp/options.py:413]: I think this should also sognify its internal state by having an underscore prefix
```suggestion
        dest='_plugin_locations', metavar='PATH', action='append',
```

**Review by Grub4K — APPROVED** [review]

File: yt_dlp/options.py (hunk 1 above)
Comment (Grub4K) [reply yt_dlp/options.py:413]: Actually, disregard this. If we plan on using the same flag (which we will), this variable and its value will not change and should be safe to consume.

Comment (bashonly) [inline yt_dlp/options.py:413]: I understand that the naming of this option was following the convention of `--ffmpeg-location` and `--config-location`, but I think following that convention is wrong to do here. Those two options allow the user to also specify the path to the **file**, not only the directory. Using the `-locations` name would imply that the user can specify the path to a wheel or zip, not just a directory. I think `--plugin-dirs` is better. It's shorter, it avoids confusion, it's more self-explanatory.

```suggestion
        '--plugin-dirs',
        dest='plugin_dirs', metavar='PATH', action='append',
```

File: yt_dlp/options.py [hunk 2]
```diff
@@ -408,6 +408,10 @@ def _alias_callback(option, opt_str, value, parser, opts, nargs):
...
+        help='Location to search for plugin packages. Can be used multiple times to add multiple directories.')
```
Comment (bashonly) [inline yt_dlp/options.py:414]: Similarly, here we should clarify that the PATH **needs** to be a directory

But more importantly: until the globals proposal is implemented, this option will only work with **extractor** plugins, and **not** postprocessor plugins (they are loaded too early). This needs to be documented in the help string for now

```suggestion
        help='Path to a directory for yt-dlp to search for plugins. Note that this currently only works for extractor plugins; postprocessor plugins can only be loaded from the default plugin directories. This option can be be used multiple times to add multiple directories')
```

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/__init__.py [hunk 3]
```diff
@@ -967,6 +969,10 @@ def _real_main(argv=None):
 
     parser, opts, all_urls, ydl_opts = parse_options(argv)
 
+    # HACK: Set the plugins location early on
+    # TODO(coletdjnz): remove when plugin globals system is implemented
+    Config._plugin_locations = opts.plugin_locations
```
Comment (bashonly) [inline yt_dlp/__init__.py:974]:
```suggestion
    Config._plugin_locations = opts.plugin_dirs
```

File: yt_dlp/__init__.py [hunk 4]
```diff
@@ -945,6 +946,7 @@ def parse_options(argv=None):
         'xattr_set_filesize': opts.xattr_set_filesize,
         'match_filter': opts.match_filter,
         'color': opts.color,
+        'plugin_locations': opts.plugin_locations,
```
Comment (bashonly) [inline yt_dlp/__init__.py:949]:
```suggestion
        'plugin_locations': opts.plugin_dirs,
```

File: yt_dlp/__init__.py [hunk 5]
```diff
@@ -945,6 +946,7 @@ def parse_options(argv=None):
         'xattr_set_filesize': opts.xattr_set_filesize,
         'match_filter': opts.match_filter,
         'color': opts.color,
+        'plugin_dirs': opts.plugin_dirs,
```
Comment (coletdjnz) [inline yt_dlp/__init__.py:949]:
```suggestion
```

as this is not a `YoutubeDL` opt

File: test/test_plugins.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 import sys
 import unittest
 from pathlib import Path
+from yt_dlp import Config
```
Comment (bashonly) [inline test/test_plugins.py:7]: should this be moved below where we import from `yt_dlp.plugins`?

File: yt_dlp/__init__.py [hunk 7]
```diff
@@ -62,6 +62,7 @@
     traverse_obj,
     variadic,
     write_string,
+    Config,
```
Comment (bashonly) [inline yt_dlp/__init__.py:65]: this needs to be moved up with the classes

File: yt_dlp/plugins.py [hunk 8]
```diff
@@ -20,6 +20,7 @@
     get_user_config_dirs,
     orderedSet,
     write_string,
+    Config,
```
Comment (bashonly) [inline yt_dlp/plugins.py:23]: same; move to top

**Review by bashonly — CHANGES_REQUESTED** [review]

File: yt_dlp/__init__.py (hunk 5 above)
Comment (imranh2) [reply yt_dlp/__init__.py]: Done

File: test/test_plugins.py (hunk 6 above)
Comment (imranh2) [reply test/test_plugins.py]: Done

File: yt_dlp/__init__.py (hunk 7 above)
Comment (imranh2) [reply yt_dlp/__init__.py]: Done

File: yt_dlp/plugins.py (hunk 8 above)
Comment (imranh2) [reply yt_dlp/plugins.py]: Done

File: yt_dlp/options.py [hunk 9]
```diff
@@ -408,6 +408,14 @@ def _alias_callback(option, opt_str, value, parser, opts, nargs):
         help=(
             'Location of the main configuration file; either the path to the config or its containing directory '
             '("-" for stdin). Can be used multiple times and inside other configuration files'))
+    general.add_option(
+        '--plugin-dirs',
+        dest='plugin_dirs', metavar='PATH', action='append',
+        help=(
+            'Path to a directory for yt-dlp to search for plugins (in addition to the default '
```
Comment (Grub4K) [inline yt_dlp/options.py:415]: I dont like this wording. How about sth like "path to an additional plugin directory"?

Comment (imranh2) [reply yt_dlp/options.py:415]: Yeah we want to make sure that people know this doesn't overwrite the defaults but appends to the list of defaults.

`Path to a directory for yt-dlp to additionally search for plugins (in addition to the default plugin directories).`

Or

`Append a path to the list of default plugin directories yt-dlp uses to search for plugins.`

Comment (imranh2) [reply yt_dlp/options.py]: Changed it to something that should be clear

File: test/test_plugins.py [hunk 10]
```diff
@@ -10,7 +10,7 @@
 sys.path.append(str(TEST_DATA_DIR))
 importlib.invalidate_caches()
 
-from yt_dlp.plugins import PACKAGE_NAME, directories, load_plugins
+from yt_dlp.plugins import Config, PACKAGE_NAME, directories, load_plugins
```
Comment (bashonly) [inline test/test_plugins.py:13]:
```suggestion
from yt_dlp import Config
from yt_dlp.plugins import PACKAGE_NAME, directories, load_plugins
```

File: test/test_plugins.py [hunk 11]
```diff
@@ -68,6 +68,24 @@ def test_importing_zipped_module(self):
             os.remove(zip_path)
             importlib.invalidate_caches()  # reset the import caches
 
+    def test_plugin_locations(self):
```
Comment (bashonly) [inline test/test_plugins.py:71]: if we're renaming the `Config` var too, then let's go all out
```suggestion
    def test_plugin_dirs(self):
```

File: test/test_plugins.py [hunk 12]
```diff
@@ -68,6 +68,24 @@ def test_importing_zipped_module(self):
...
+        # Internal plugin locations hack for CLI --plugin-dirs
+        # To be replaced with proper system later
+        custom_plugin_locations_path = TEST_DATA_DIR / 'plugin_packages'
+        Config._plugin_dirs = [str(custom_plugin_locations_path)]
```
Comment (bashonly) [inline test/test_plugins.py:75]:
```suggestion
        # Internal plugin dirs hack for CLI --plugin-dirs
        # To be replaced with proper system later
        custom_plugin_dir = TEST_DATA_DIR / 'plugin_packages'
        Config._plugin_dirs = [str(custom_plugin_dir)]
```

File: test/test_plugins.py [hunk 13]
```diff
@@ -68,6 +68,24 @@ def test_importing_zipped_module(self):
...
+        importlib.invalidate_caches()  # reset the import caches
+
+        try:
+            package = importlib.import_module(f'{PACKAGE_NAME}.extractor')
+            self.assertIn(custom_plugin_locations_path / 'testpackage' / PACKAGE_NAME / 'extractor', map(Path, package.__path__))
```
Comment (bashonly) [inline test/test_plugins.py:80]:
```suggestion
            self.assertIn(custom_plugin_dir / 'testpackage' / PACKAGE_NAME / 'extractor', map(Path, package.__path__))
```

File: yt_dlp/__init__.py [hunk 14]
```diff
@@ -967,6 +968,10 @@ def _real_main(argv=None):
 
     parser, opts, all_urls, ydl_opts = parse_options(argv)
 
+    # HACK: Set the plugins location early on
```
Comment (bashonly) [inline yt_dlp/__init__.py:971]:
```suggestion
    # HACK: Set the plugin dirs early on
```

File: yt_dlp/options.py [hunk 15]
```diff
@@ -408,6 +408,14 @@ def _alias_callback(option, opt_str, value, parser, opts, nargs):
...
             'Location of the main configuration file; either the path to the config or its containing directory '
             '("-" for stdin). Can be used multiple times and inside other configuration files'))
+    general.add_option(
+        '--plugin-dirs',
+        dest='plugin_dirs', metavar='PATH', action='append',
+        help=(
+            'Path to append to the default list of directories yt-dlp uses to search for plugins. '
+            'Note that this currently only works for extractor plugins; '
+            'postprocessor plugins can only be loaded from the default plugin directories. This '
+            'option can be used multiple times to add multiple directories'))
```
Comment (bashonly) [inline yt_dlp/options.py:418]: I liked Grub's suggestion better
```suggestion
        help=(
            'Path to an additional directory to search for plugins. '
            'Note that this currently only works for extractor plugins; '
            'postprocessor plugins can only be loaded from the default plugin directories. '
            'This option can be used multiple times to add multiple directories'))
```

File: README.md [hunk 16]
```diff
@@ -348,6 +348,13 @@ If you fork the project on GitHub, you can run your fork's [build workflow](.git
                                     containing directory ("-" for stdin). Can be
                                     used multiple times and inside other
                                     configuration files
+    --plugin-dirs PATH              Path to a directory for yt-dlp to search for
+                                    plugins (in addition to the default plugin
+                                    directories). Note that this currently only
+                                    works for extractor plugins; postprocessor
+                                    plugins can only be loaded from the default
+                                    plugin directories. This option can be used
+                                    multiple times to add multiple directories
```
Comment (bashonly) [inline README.md:357]: need to run `make README.md` again (do this last)

**Review by bashonly — CHANGES_REQUESTED** [review]

**Review by bashonly — APPROVED** [review]

File: test/test_plugins.py [hunk 17]
```diff
@@ -10,6 +10,7 @@
 sys.path.append(str(TEST_DATA_DIR))
 importlib.invalidate_caches()
 
+from yt_dlp import Config
```
Comment (bashonly) [inline test/test_plugins.py:13]:
```suggestion
from yt_dlp.utils import Config
```

File: yt_dlp/options.py [hunk 18]
```diff
@@ -408,6 +408,14 @@ def _alias_callback(option, opt_str, value, parser, opts, nargs):
...
             'Location of the main configuration file; either the path to the config or its containing directory '
             '("-" for stdin). Can be used multiple times and inside other configuration files'))
+    general.add_option(
+        '--plugin-dirs',
+        dest='plugin_dirs', metavar='PATH', action='append',
+        help=(
+            'Path to an additional directory to search for plugins. '
+            'Note that this currently only works for extractor plugins; '
+            'postprocessor plugins can only be loaded from the default plugin directories. '
+            'This option can be used multiple times to add multiple directories'))
```
Comment (bashonly) [inline yt_dlp/options.py:418]:
```suggestion
            'This option can be used multiple times to add multiple directories. '
            'Note that this currently only works for extractor plugins; '
            'postprocessor plugins can only be loaded from the default plugin directories'))
```

**Review by Grub4K — APPROVED** [review]

**Pull request merged**